tidy3d/plugins/design/design.py

```diff
@@ -102,22 +128,190 @@ def run(self, function: Callable, **kwargs) -> Result:
+        )
+
+    def run_single(self, fn: Callable, console) -> Tuple(list[dict], list, list[Any]):
+        """Run a single function of parameter inputs."""
+        evaluate_fn = self._get_evaluate_fn_single(fn=fn)
+        return self.method.run(run_fn=evaluate_fn, parameters=self.parameters, console=console)
+
+    def run_pre_post(self, fn_pre: Callable, fn_post: Callable, console) -> Tuple(
+        list[dict], list[dict], list[Any]
+    ):
+        """Run a function with tidy3d implicitly called in between."""
+        handler = self._get_evaluate_fn_pre_post(
+            fn_pre=fn_pre, fn_post=fn_post, fn_mid=self._fn_mid
+        )
+        fn_args, fn_values, aux_values, opt_output = self.method.run(
+            run_fn=handler.evaluate, parameters=self.parameters, console=console
+        )
+        return fn_args, fn_values, aux_values, opt_output, handler.sim_names
+
+    """ Helpers """
+
+    def _get_evaluate_fn_single(self, fn: Callable) -> list[Any]:
+        """Get function that sequentially evaluates single `fn` for a list of arguments."""
+
+        def evaluate(args_list: list) -> list[Any]:
+            """Evaluate a list of arguments passed to ``fn``."""
+            return [fn(**args) for args in args_list]
+
+        return evaluate
+
+    def _get_evaluate_fn_pre_post(self, fn_pre: Callable, fn_post: Callable, fn_mid) -> list[Any]:
+        """Get function that tries to use batch processing on a set of arguments."""
+
+        class Pre_Post_Handler:
+            def __init__(self):
+                self.sim_counter = 0
+                self.sim_names = []
+
+            def _fn_combined(self, args_list):
+                """Compute fn_pre and fn_post functions and capture other outputs."""
+                sim_dict = {str(idx): fn_pre(**arg_list) for idx, arg_list in enumerate(args_list)}
+                data, sim_names = fn_mid(sim_dict, self.sim_counter)
+                self.sim_names.extend(sim_names)
+                self.sim_counter += len(sim_names)
+                post_out = [fn_post(val[1]) for val in data.items()]
+                return post_out
+
+            def evaluate(self, args_list: list[tuple[Any, ...]]) -> list[Any]:
+                """Put together into one pipeline."""
+                out = self._fn_combined(args_list)
+                return out
+
+        handler = Pre_Post_Handler()
+
+        return handler
+
+    def _fn_mid(
+        self, pre_out: dict[int, Any], sim_counter: int
+    ) -> Union[dict[int, Any], BatchData]:
+        """A function of the output of ``fn_pre`` that gives the input to ``fn_post``."""
+
+        # NOTE: Need to check that pre_out has string keys!
+
+        # Keep copy of original to use if no tidy3d simulation required
+        original_pre_out = pre_out.copy()
+
+        # Convert list input to dict of dict before passing through checks
+        was_list = False
+        if all(isinstance(sim, list) for sim in pre_out.values()):
```

Comment:
Good idea. I've checked now and it's not possible for one of `pre_out.values` to be a list and the other to be anything else. `pre_out.values` is the return from the user supplied fn_pre which doesn't change during the run. If the return for fn_pre were say `return list, string` it would error before reaching this check.

Reply:
> If the return for fn_pre were say return list, string it would error before reaching this check.

I'm a bit confused, wouldn't `list, string` be acceptable?

Reply:
So played around with this again and now it fails with a proper raise. The `list, string` example would be a tuple output from fn_pre which is not an accepted output. 

If it were `[list, string]` this would fail as well as the second element from fn_pre needs to be a list or dict.